src/dataflow/src/render/top_k.rs

```diff
@@ -62,62 +62,67 @@ where
                 collection
                     .map(move |((key, hash), row)| ((key, hash % modulus), row))
                     .reduce_named("TopK", {
-                        move |_key, source, target| {
-                            target.extend(source.iter().map(|&(row, diff)| (row.clone(), diff)));
+                        move |_key, source, target: &mut Vec<(Row, isize)>| {
+                            // Determine if we must actually shrink the result set.
                             let must_shrink = offset > 0
                                 || limit
                                     .map(|l| {
-                                        target.iter().map(|(_, d)| *d).sum::<isize>() as usize > l
+                                        source.iter().map(|(_, d)| *d).sum::<isize>() as usize > l
                                     })
                                     .unwrap_or(false);
                             if must_shrink {
+                                // local copies that may count down to zero.
+                                let mut offset = offset;
+                                let mut limit = limit;
+
+                                // The order in which we should produce rows.
+                                let mut indexes = (0..source.len()).collect::<Vec<_>>();
                                 if !order_clone.is_empty() {
+                                    // We decode the datums once, into a common buffer for efficiency.
+                                    // Each row should contain an identical number of columns; we should check that.
+                                    // The alternative is `Vec<Vec<Datum>>`, which involves substantially more allocations.
+                                    let mut buffer = Vec::new();
+                                    for row in source.iter() {
+                                        buffer.extend(row.0.iter());
```

Comment:
After unpacking the first row, we know the correct size of `buffer`, right? Should we call `reserve` ?

Reply:
In fact, we should be able to get it as input and assert that each decoded row has the right length! I'll do that now.